### DISCUSSION TITLE
Accessing the request URI in Authenticator fails at runtime

### DISCUSSION THREAD

**Discussion by CollinAlpert** [keycloak/keycloak#10628] [Q&A]
I have a custom Authenticator SPI and in the `authenticate` method I need to access the request URI. If I call `context.getHttpRequest().getUri()`, the following exception is thrown when the Authenticator is run:
```
Uncaught server error: 
java.lang.NoSuchMethodError: 'org.jboss.resteasy.spi.ResteasyUriInfo org.jboss.resteasy.spi.HttpRequest.getUri()'
```

These are my dependencies:
```xml
<dependency>
    <groupId>org.keycloak</groupId>
    <artifactId>keycloak-services</artifactId>
    <version>${keycloak.version}</version>
    <scope>provided</scope>
</dependency>
<dependency>
    <groupId>org.keycloak</groupId>
    <artifactId>keycloak-server-spi</artifactId>
    <version>${keycloak.version}</version>
    <scope>provided</scope>
</dependency>
<dependency>
    <groupId>org.keycloak</groupId>
    <artifactId>keycloak-server-spi-private</artifactId>
    <version>${keycloak.version}</version>
    <scope>provided</scope>
</dependency>
```

What am I missing? I am using Keycloak 17 (both in server and Maven dependencies)

**Comment by ariel-adler** [comment]
It also happens in version 16.1.1
Any idea?

**Reply by CollinAlpert** [reply]
I have found it to be related to #9599 (specifically [this comment](https://github.com/keycloak/keycloak/issues/9599#issuecomment-1016509074)). You can see if one of the solutions posted there will help you.

**Reply by ariel-adler** [reply]
Thank you Collin it works :)

Add this to maven:

         <properties>
             <resteasy.version>4.7.4.Final</resteasy.version>
        </properties>

        <dependency>
            <groupId>org.jboss.resteasy</groupId>
            <artifactId>resteasy-client</artifactId>
            <version>${resteasy.version}</version>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>org.jboss.resteasy</groupId>
            <artifactId>resteasy-jackson2-provider</artifactId>
            <version>${resteasy.version}</version>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>org.jboss.resteasy</groupId>
            <artifactId>resteasy-multipart-provider</artifactId>
            <version>${resteasy.version}</version>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>org.jboss.resteasy</groupId>
            <artifactId>resteasy-jaxb-provider</artifactId>
            <version>${resteasy.version}</version>
            <scope>provided</scope>
        </dependency>

**Reply by CollinAlpert** [reply]
Actually I have found that adding only `resteasy-multipart-provider` works as well.

**Reply by ariel-adler** [reply]
Good to know.
Thank you!

**Comment by fzoli** [comment]
For me the provided dependency does not solve the problem.
I build a flat JAR with Gradle. In this build system provided means that the library is used only at compile time, but it is not part of the flat JAR.

Anyway I found a workaround:

I replaced `context.getHttpRequest().getUri().getQueryParameters()` to `context.getSession().getContext().getUri().getDelegate().getQueryParameters()`

**Reply by jamesdh** [reply]
We're on version 15 still and we started seeing this recently. Not entirely sure where the dependency change came from since we haven't touched anything. In any case, @fzoli's solution appears to work regardless of the version of resteasy that is used.
